From: Kemeng Shi <shikemeng@huaweicloud.com>
Subject: mm: swap: avoid losing cluster in swap_reclaim_full_clusters()
Date: Sun, 23 Feb 2025 00:08:45 +0800

If no swap cache is reclaimed, cluster taken off from full_clusters list
will not be put in any list and may not be reused.  Do relocate_cluster
for such cluster to fix the issue.

Link: https://lkml.kernel.org/r/20250222160850.505274-2-shikemeng@huaweicloud.com
Fixes: 3b644773eefda ("mm, swap: reduce contention on device lock")
Signed-off-by: Kemeng Shi <shikemeng@huaweicloud.com>
Reviewed-by: Kairui Song <kasong@tencent.com>
Signed-off-by: Andrew Morton <akpm@linux-foundation.org>
---

 mm/swapfile.c |    4 ++++
 1 file changed, 4 insertions(+)

--- a/mm/swapfile.c~mm-swap-avoid-losting-cluster-in-swap_reclaim_full_clusters
+++ a/mm/swapfile.c
@@ -858,6 +858,10 @@ static void swap_reclaim_full_clusters(s
 			offset++;
 		}
 
+		/* in case no swap cache is reclaimed */
+		if (ci->flags == CLUSTER_FLAG_NONE)
+			relocate_cluster(si, ci);
+
 		unlock_cluster(ci);
 		if (to_scan <= 0)
 			break;
_
